feat: Add top-level parent logic to Upstream/Dowstream links [FC-0097]

ChrisChV opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

- Adds the `top_level_parent_usage_key` to the `EntityLinkBase`
- This field is used to save the top-level parent of a component or container when it is imported into a course. Example: A unit with components imported into a course. The unit is the top-level parent of the components.
- Updates the `DownstreamListView` to return the top-level parents instead of downstream child, if this parent exists.
- Each time containers with children were synchronized, a new downstream block was created for each child instead of updating the existing one. This occurred because the `upstream_key` was incorrectly validated as an `Opaquekey` against a list of key strings. This was fixed by converting the `upstream_key` to a string before the verification. (see https://github.com/openedx/edx-platform/pull/37076/commits/34cd5a4781826f2439dce0ca93edca49b9f0353f and https://github.com/openedx/edx-platform/pull/37076/commits/29647831dc3142afeb4fdcef85ce1c2c4d8e2e77)
- Which edX user roles will this change impact?  "Course Author", "Developer".


## Supporting information

- Github Issue: https://github.com/openedx/frontend-app-authoring/issues/2168#issuecomment-3097415384
- Used in: https://github.com/openedx/frontend-app-authoring/pull/2271
- Internal ticket: [FAL-4231](https://tasks.opencraft.com/browse/FAL-4231)

## Testing instructions

- Verify that the test covers the new code.
- Follow the testing instructions in https://github.com/openedx/frontend-app-authoring/pull/2271

## Deadline

None

## Other information

- TODO: It's missing some changes in the counts of the summary (see https://github.com/openedx/frontend-app-authoring/issues/2168#issuecomment-3129284830)

ChrisChV committed:
feat: Add top_level_parent field and updates API to optain the top lavel parent in an update

openedx-webhooks commented:
Thanks for the pull request, @ChrisChV!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

ChrisChV committed:
test: Added tests for top-level parents logic

ChrisChV committed:
style: Fix broken lint

navinkarkera commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -175,6 +188,26 @@ def filter_links(
+        if use_top_level_parents is not None:
+            # Get objects without top_level_parent_usage_key
+            non_top_level_objects = result.filter(top_level_parent_usage_key=UsageKeyField.Empty)
+            non_top_level_ids = non_top_level_objects.values_list('id', flat=True)
```
`non_top_level_objects` is confusing, the name indicates that these objects are not at top level, but the comment says otherwise.

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/37076/commits/bc484ec4b1bb88047cd8782f281594470ae1bc67

navinkarkera commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -175,6 +188,26 @@ def filter_links(
+
+            top_level_objects = ContainerLink.filter_links(**{
+                "downstream_usage_key__in": top_level_keys
+            })
```
Same here, `top_level_*` variables should be the ones that are at the top, but here they are the objects that are under some other container.

ChrisChV replied:
Not really. The `top_level_parent` always represents the highest-level parent at import time, not the direct parent. This code retrieves the top-level parent keys and their objects (containers) and returns them instead of their descendant.

navinkarkera commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -175,6 +188,26 @@ def filter_links(
+            non_top_level_ids = non_top_level_objects.values_list('id', flat=True)
+
+            # Get the non-null top_level_parent_usage_key
+            # Note: using `exclude()` raise `TypeError`
```
Not sure if I understand this comment, you seem to be using `exclude`.

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/37076/commits/bc484ec4b1bb88047cd8782f281594470ae1bc67

navinkarkera commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -308,13 +343,22 @@ def filter_links(
+        `use_top_level_parents` is an special filter, replace any result with the top-level parent if exists.
+        Example: We have linkA and linkB with top-level parent as linkC and linkD without top-level parent.
+        After all other filters:
+        Case 1: `use_top_level_parents` is False, the result is [linkA, linkB, linkD]
```
```suggestion
        Case 1: `use_top_level_parents` is False, the result is [linkA, linkB, linkC, linkD]
```

It should list all components no?

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/37076/commits/bc484ec4b1bb88047cd8782f281594470ae1bc67

navinkarkera replied:
@ChrisChV Just want to confirm whether `linkC` will not show up if `use_top_level_parents` is False?

navinkarkera commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -327,6 +371,36 @@ def filter_links(
+                        Coalesce("version_declined", 0)
+                    )
+                )
+            )
```
With help of some AI and django ORM, we can simplify this as below. The same can be done with `ComponentLinks.filter_links` function.

```suggestion
        # Handle top-level parents logic
        if use_top_level_parents is not None:
            # Subquery to find top-level parent usage keys
            top_level_parent_subquery = cls.objects.filter(
                downstream_usage_key=OuterRef('top_level_parent_usage_key')
            ).values('id')

            # Annotate with a flag indicating if the object has a top-level parent
            result = result.annotate(
                has_top_level_parent=Exists(top_level_parent_subquery)
            )

            # If use_top_level_parents is True, filter to get only top-level parents or objects without parents
            return result.filter(
                Q(has_top_level_parent=True) | Q(top_level_parent_usage_key=UsageKeyField.Empty)
            )
```

ChrisChV replied:
This code does not replace descendants with their top-level parents. See https://github.com/openedx/edx-platform/pull/37076#discussion_r2240671472

I have tried using the `top_level_parent_subquery` in my code, but it has become complicated. Do you think the code needs major changes?

navinkarkera replied:
Ahh, that is good then. We could however reduce a call to database by annotating `top_level_objects` query with `ready_to_sync` and skip the final result call, something like below **(Note: you will have to shuffle this [list](https://github.com/open-craft/edx-platform/blob/806cb4f3f1260905b3306d03dbf86dd85786dc82/cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstreams.py#L1125) in test with the last one to pass.)**

```python
            top_level_objects = cls.objects.filter(
                downstream_usage_key__in=top_level_keys,
            ).select_related(*RELATED_FIELDS).annotate(
                ready_to_sync=(
                    GreaterThan(
                        Coalesce("upstream_container__publishable_entity__published__version__version_num", 0),
                        Coalesce("version_synced", 0)
                    ) & GreaterThan(
                        Coalesce("upstream_container__publishable_entity__published__version__version_num", 0),
                        Coalesce("version_declined", 0)
                    )
                )
            )

            result = list(chain(top_level_objects, non_top_level_objects))
```

And it would be nice to create a function that annotates the query with `ready_to_sync` field as it seems to be repeated here.

ChrisChV replied:
Thanks! Updated: https://github.com/openedx/edx-platform/pull/37076/commits/966c998b2d24d564cd1d17eb3cb238034979d809

navinkarkera commented on cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstreams.py:
```diff
@@ -820,139 +1056,91 @@ def test_200_all_component_downstreams_for_a_course(self):
+        )
         assert response.status_code == 200
         data = response.json()
+        self.assertEqual(data["count"], 3)
```
After making the above suggested change, this increases to 4, which seems correct as we are also making a change to top level subsection in line 1096 no?

ChrisChV replied:
Not really. According to the requirements, only the top-level parents should be displayed. Even if all containers and components within a section are updated, only the top-level parent, which is the section, should be displayed. In this test, the top-level parent of the subsection and component is the section. This test checks that only the section is displayed and is not duplicated in the result.

The test fails because the code above is not replacing descendants with their top-level parent, it is adding the subsection to the response.

I added another more simple tests for this: https://github.com/openedx/edx-platform/pull/37076/commits/bc484ec4b1bb88047cd8782f281594470ae1bc67#diff-cdab0b43d246b3a18252af57261eefce305c7a0ed93bce4549288127324c5bf3R1085

navinkarkera reviewed (changes requested):
@ChrisChV Nice work! This works as expected but I have left some suggestions and a doubt in the tests.

ChrisChV committed:
style: Nits on the code

ChrisChV committed:
style: fix broken lints

navinkarkera commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -327,6 +374,37 @@ def filter_links(
+            non_top_level_ids = non_top_level_objects.values_list('id', flat=True)
+
+            # Get the non-null top_level_parent_usage_key
+            # Note: using `exclude()` raise `TypeError`
```
```suggestion
            # Note: using `exclude(top_level_parent_usage_key=UsageKeyField.Empty)` raise `TypeError`
```

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/37076/commits/966c998b2d24d564cd1d17eb3cb238034979d809

navinkarkera reviewed (approved):
@ChrisChV Great work, it is not easy to catch all the edge cases here! :+1:
- [x] I tested this: (Played around with multiple containers from library in courses)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

navinkarkera commented:
@ChrisChV One more request, could you update the `process_list` function [here](https://github.com/open-craft/edx-platform/blob/806cb4f3f1260905b3306d03dbf86dd85786dc82/cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py#L365) to something like this:

```python
if item["last_published_at"] and merged[key]["last_published_at"] and item["last_published_at"] > merged[key]["last_published_at"]:
```

I was testing https://github.com/openedx/edx-platform/pull/37008 from @rpenido and I think it would be possible to copy unpublished blocks and this line fails in that case.

OR we should not allow users to copy unpublished blocks from library,

ChrisChV commented:
> I was testing #37008 from @rpenido and I think it would be possible to copy unpublished blocks and this line fails in that case.
> 
> OR we should not allow users to copy unpublished blocks from library,

It's a good question. I agree with not allowing users to copy unpublished blocks from libraries. @edschema @sdaitzman What do you think?

rpenido commented:
> OR we should not allow users to copy unpublished blocks from library,

For more context, when we copy/paste a component from a library, we are copying the latest draft version and linking it.
When we use the "Library Content" to add a component, we only show and add the published version of the element.

ChrisChV committed:
feat: Add _annotate_query_with_ready_to_sync to ContainerLink

ChrisChV committed:
Merge branch 'master' into chris/FAL-4231-top-level-parent

ChrisChV committed:
fix: Bug with creating blocks on every sync

Each time containers with children were synchronized, a new downstream block was created for each child
instead of updating the existing one. This occurred because the upstream_key was incorrectly validated as
an Opaquekey against a list of key strings. This was fixed by converting the upstream_key to a string before the verification.

ChrisChV committed:
test: Add tests to verify the fix of the bug of creating blocks on every sync

edschema commented:
> > OR we should not allow users to copy unpublished blocks from library,
> 
> For more context, when we copy/paste a component from a library, we are copying the latest draft version and linking it. When we use the "Library Content" to add a component, we only show and add the published version of the element.

Hi @rpenido @ChrisChV and team, thanks for catching this! Users should not be able to reuse unpublished blocks. This is a good case I don't believe we have accounted for. I made an issue to track disabling copy from the context menu for unpublished blocks here: https://github.com/openedx/frontend-app-authoring/issues/2344

If copying a block from a library that has been published but has draft changes, when pasted, that block should show the latest published version in the course. By this I mean:
- Block C is published
- Block C is edited to Block C', but Block C' is not published
- If a user copies Block C' and pastes into their course, the version pasted should be Block C

ChrisChV committed:
style: Fix broken lint

rpenido commented:
Thanks @edschema! I left a comment on the new issue.

bradenmacdonald commented:
> Adds the top_level_parent_usage_key to the EntityLinkBase
> This field is used to save the top-level parent of a component or container when it is imported into a course. Example: A unit with components imported into a course. The unit is the top-level parent of the components.

I'm curious, could you achieve the same effect with an `is_top_level` boolean field, or is it necessary to store the full ID of the top level entity? (edit: Oh, I see - we don't really have the hierarchy available when we're querying the entity links table, so without the full ID we can't merge changes from children into the top level parent item. Makes sense.)

Also, when we get to unlinking (https://github.com/openedx/frontend-app-authoring/issues/2164) we'll need to account for recursively updating the `top_level_parent_usage_key`. I'll add a note to that issue.

bradenmacdonald commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -98,6 +97,11 @@ class EntityLinkBase(models.Model):
+    # importing a container that has one or more levels of children.
+    # This represents the parent (container) in the top level
+    # at the moment of the import.
+    top_level_parent_usage_key = UsageKeyField(max_length=255, null=True, blank=True, db_index=True)
```
Do we need `null=True`? Django convention is to [never use `NULL` for string fields](https://docs.djangoproject.com/en/5.2/ref/models/fields/#django.db.models.Field.null).

bradenmacdonald replied:
Also, did you consider making this a `ForeignKey` to `ContainerLink` instead of a string ID? That would help it stay up to date more consistently. (We wanted _everything_ in these models to be a foreign key where possible, but we can't until we've converted courses to use Learning Core too)

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/37076/commits/633e0500ba0d53476f7df714903ce3843d3c2193

bradenmacdonald commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -153,16 +157,25 @@ def upstream_context_title(self) -> str:
         """
         Get all links along with sync flag, upstream context title and version, with optional filtering.
+
+        `use_top_level_parents` is an special filter, replace any result with the top-level parent if exists.
+        Example: We have linkA and linkB with top-level parent as linkC, and linkD without top-level parent.
+        After all other filters:
+        Case 1: `use_top_level_parents` is False, the result is [linkA, linkB, linkC, linkD]
+        Case 2: `use_top_level_parents` is True, the result is [linkC, linkD]
```
I'm not a big fan of hiding this "special filter" in the `link_filter` kwargs  (because it's really a boolean parameter for the function - it does more than just remove some results; it replaces them). I think it will be cleaner if you just make it an explicit parameter like this:

```python
def filter_links(cls, *, use_top_level_parents=False, **link_filter):
```
That should also be fully backwards-compatible.

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/37076/commits/e77a67f214681adbbb09a5a58e07b8a1c2800e57

bradenmacdonald commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -308,13 +346,49 @@ def filter_links(
+        ]
+
+        ready_to_sync = link_filter.pop('ready_to_sync', None)
+        use_top_level_parents = link_filter.pop('use_top_level_parents', None)
```
Same minor comment here about making this an explicit separate parameter instead of putting it into `link_filter`.

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/37076/commits/e77a67f214681adbbb09a5a58e07b8a1c2800e57

bradenmacdonald commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -175,6 +188,29 @@ def filter_links(
+        # Handle top-level parents logic
+        if use_top_level_parents is not None:
+            # Get objects without top_level_parent_usage_key
+            objects_without_top_level = result.filter(top_level_parent_usage_key=UsageKeyField.Empty)
```
Above this variable is called `non_top_level_objects` and here it's called `objects_without_top_level`.
Same for `non_top_level_ids` and `ids_without_top_level`.
Would it make sense to make those consistent? And maybe adding a comment in each stating that this logic is very similar to what's in that other function, but pointing out the differences (if any)

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/37076/commits/e77a67f214681adbbb09a5a58e07b8a1c2800e57

bradenmacdonald commented on cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstream_sync_integration.py:
```diff
@@ -260,6 +295,7 @@ def test_unit_sync(self):
                     editor="visual"
                     upstream="{self.upstream_html1['id']}"
                     upstream_version="2"
+                    top_level_downstream_parent="{downstream_unit["locator"]}"
```
This will work for copy-paste and other use cases, but course exports might be a problem. When you export the course, you can import it with a different ID. and then the `top_level_downstream_parent` IDs will all be wrong because they include the course ID part which has now changed.

We could store just the [top level parent block type, top level parent block ID] instead, which would be stable even if courses are re-run, imported, duplicated, etc.

OR instead of making `top_level_downstream_parent` a usagekey field, we could make it a foreign key to the parent *Link model, and then in the OLX just store is_top_level=True/False.

Actually now I'm wondering if we even have the logic to duplicate all the ContainerLink/ComponentLink objects when you re-run a course? Might be something we have overlooked?

navinkarkera replied:
I don't think we are exporting the links along with course at all, we also don't duplicate them on re-run. Simplest option to fix it would be to run [recreate_upstream_links](https://github.com/openedx/edx-platform/blob/c9886882e3280b84fed1bf87d01a74e745e88d75/cms/djangoapps/contentstore/management/commands/recreate_upstream_links.py#L94) for the specific course after importing a course or re-run.

bradenmacdonald replied:
> Simplest option to fix it would be to run [recreate_upstream_links](https://github.com/openedx/edx-platform/blob/c9886882e3280b84fed1bf87d01a74e745e88d75/cms/djangoapps/contentstore/management/commands/recreate_upstream_links.py#L94) for the specific course after importing a course or re-run.

Good point. I created https://github.com/openedx/frontend-app-authoring/issues/2269 for this. @ChrisChV  can you please review?

bradenmacdonald commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -327,6 +327,17 @@ class UpstreamSyncMixin(XBlockMixin):
 
+    top_level_downstream_parent = String(
+        help=(
+            "The usage key of the downstream block that is the top-level parent of "
```
See my other comment above. I'm worried about encoding full usage keys in XBlock fields. I think there is a [`Reference` field type](https://github.com/openedx/XBlock/blob/8c3599e02c902b79a9e3e5c5e800b46666202b1a/xblock/fields.py#L1023) that is supposed to handle this properly (store the type+ID but not the course ID part), but I'm not sure if it's actually used anywhere or if it works.

ChrisChV replied:
Updated here https://github.com/openedx/edx-platform/pull/37076/commits/49e1989725c8d9a7c125d3417143576b1c25c25b

I have used a simple method to save a string "block_type@block_id". Neither the `Reference` class nor the `DefinitionKey` class helped me much. 

> OR instead of making top_level_downstream_parent a usagekey field, we could make it a foreign key to the parent *Link model, and then in the OLX just store is_top_level=True/False.

Regarding using a boolean, a descendant wouldn't know exactly what its top-level parent is; the boolean doesn't give that information.

bradenmacdonald replied:
Your simple string method sounds fine to me, _but_ I don't think you should call it a "definition key". In split modulestore, definition refers to a very specific thing that contains a subset of the XBlock fields, and I think it's going to be confusing if you call this a "definition key".

I guess you are thinking of [`DefinitionLocator`](https://github.com/openedx/opaque-keys/blob/1af94ae52c42d5cf03f6df979147098fc205c114/opaque_keys/edx/locator.py#L1195-L1226) and its related `DefinitionKey` which _does_ represent a pair of [block_type, ID] but the ID is not a block_id - it's a mongoDB object ID. So storing a [block_type, block_id] and calling it a "definition key" (which is always a [block_type, mongoDB object ID]) is going to be confusing.

There is actually a thing you can use for this though - called `BlockKey`:

https://github.com/openedx/edx-platform/blob/636ab6f9cf1163449c7b62313fdf056e7cbbfa06/xmodule/util/keys.py#L13-L29

I think you can keep all your code the same, but just rename `top_level_downstream_parent_def` to `top_level_downstream_parent_key` and maybe rewrite `get_definition_from_usage_key` to use `BlockKey.from_usage_key()` before converting it to a string .

ChrisChV replied:
@bradenmacdonald Thanks! It works

Updated in: https://github.com/openedx/edx-platform/pull/37076/commits/c664a7afcb653ab7866b1eaa520c252c81309b82

bradenmacdonald reviewed (commented):
Very nice work @ChrisChV! I think this is great, but I do have some concerns about import-export and re-runs, because of the way you're storing full usage IDs. I wonder if we can achieve the same thing without storing the course ID part.

ChrisChV committed:
refactor: Set use_top_level_parents as special filter separated from link_filter

ChrisChV committed:
refactor: top_level_parent to ForeignKey of ContainerLink

ChrisChV committed:
refactor: Use definition of the block instead of the full key in top-level parent

ChrisChV committed:
style: Fix broken lin

ChrisChV committed:
Merge branch 'master' into chris/FAL-4231-top-level-parent

ChrisChV commented:
@bradenmacdonald  This and https://github.com/openedx/frontend-app-authoring/pull/2271 are ready for another review

I will do this https://github.com/openedx/frontend-app-authoring/issues/2168#issuecomment-3144686429 in a follow-up PR. I would like to merge this PR to unlock other tasks.

bradenmacdonald reviewed (commented):
Thanks @ChrisChV ! Looking good, just one concern about the name "definition" and then once you've renamed that and @navinkarkera confirms he's tested the final version, I will approve.

ChrisChV committed:
Merge branch 'master' into chris/FAL-4231-top-level-parent

ChrisChV committed:
refactor: Use BlockType in top_level_downstream_parent_key

ChrisChV commented:
> Thanks @ChrisChV ! Looking good, just one concern about the name "definition" and then once you've renamed that and @navinkarkera confirms he's tested the final version, I will approve.

@navinkarkera It's ready for another review

ChrisChV committed:
style: Fix broken lint

ChrisChV commented:
@bradenmacdonald @navinkarkera I need some help here.

These [tests](https://github.com/openedx/edx-platform/actions/runs/16842111633/job/47715197815?pr=37076) are failing because of this change https://github.com/openedx/edx-platform/pull/37076/commits/633e0500ba0d53476f7df714903ce3843d3c2193#diff-5b6284478c70bbd62aaaf2b39e1d5b4b526214ecf8aea0cea9fefc84fdab79e7R277

Adding the `delay` solves the problem, but it is not what is needed now. It seems that some configuration is missing in the tests, but I can't find the solution.

bradenmacdonald commented:
@ChrisChV  Are you talking about this?

https://github.com/openedx/edx-platform/blob/633e0500ba0d53476f7df714903ce3843d3c2193/cms/djangoapps/contentstore/signals/handlers.py#L273-L277

What I would suggest is that in the child handler, when you go to create the link and find that the parent link doesn't exist yet, then create the parent link synchronously and idempotently (call `handle_create_or_update_xblock_upstream_link ` for the parent from within the child `handle_create_or_update_xblock_upstream_link ` handler).  Later, the handler will get called to create the parent link and it will already exist, so it will be a no-op, but that's fine. Would something like that work?

ChrisChV committed:
refactor: Verify creation of top-level parents

ChrisChV committed:
Merge branch 'master' into chris/FAL-4231-top-level-parent

ChrisChV commented:
@bradenmacdonald It works! Thanks! Those changes are ready for review if you wish to take a look: https://github.com/openedx/edx-platform/pull/37076/commits/6cc840ac00acbca068640e3347805f06a87e0a2d

> Thanks @ChrisChV ! Looking good, just one concern about the name "definition" and then once you've renamed that and @navinkarkera confirms he's tested the final version, I will approve.

@navinkarkera It's ready for a final review

bradenmacdonald commented:
@ChrisChV Glad to hear! I'll let @navinkarkera do a final check and then you can go ahead and merge. If he's not able to review by tomorrow though let me know.

navinkarkera reviewed (approved):
@ChrisChV Nice work! Tested this locally with subsections and units.

ChrisChV committed:
style: Update comment

ChrisChV commented:
> Just want to confirm whether linkC will not show up if use_top_level_parents is False?

Yes, I updated the comment. Thanks! https://github.com/openedx/edx-platform/pull/37076/commits/5d806b000d65248bc260eb17f49b83af36eadbcd

ChrisChV committed:
Merge branch 'master' into chris/FAL-4231-top-level-parent

ChrisChV committed:
Merge branch 'master' into chris/FAL-4231-top-level-parent

kdmccormick commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -327,6 +327,17 @@ class UpstreamSyncMixin(XBlockMixin):
         default=None, scope=Scope.settings, hidden=True, enforce_type=True,
     )
 
+    top_level_downstream_parent_key = Dict(
+        help=(
+            "The block key ('block_type@block_id') of the downstream block that is the top-level parent of "
+            "this block. This is present if the creation of this block is a consequence of "
+            "importing a container that has one or more levels of children. "
+            "This represents the parent (container) in the top level "
+            "at the moment of the import."
+        ),
+        default=None, scope=Scope.settings, hidden=True, enforce_type=True,
+    )
```
Dicts serialize to OLX in a very hard-to-read way:
```xml
<problem top_level_downstream_parent_key="{&quot;type&quot;: &quot;html&quot;, &quot;id&quot;: &quot;my-id&quot;}">...</problem>
```

This field could instead be a String, joined with a `:`, like this:
```xml
<problem top_level_downstream_parent_key="html:my-id">...</problem>
```

@ChrisChV @bradenmacdonald Do you think it'd be worth making this change now? After the Ulmo cutoff, we're locked in to the Dict field.

bradenmacdonald replied:
Yes I would much prefer something like that.

navinkarkera replied:
@kdmccormick @bradenmacdonald Created a PR: https://github.com/openedx/edx-platform/pull/37448 which converts this field to a string in this format `html@my-id`.
